### ISSUE TITLE
It's not possible to run tests from a repo like sdk using the JSON reporter

### ISSUE THREAD

**Issue by DanTup** [dart-lang/sdk#33466]
Splitting this from #32952 since it was hi-jacking an unrelated thread.

I'm working on a test runner for VS Code which relies on the JSON reporter from the `test` package. The JSON reporter is only available when running tests via `pub run test` and not when running the test script directly through `dart`. This seems by design and difficult to change.

The tests in the SDK repo (for example in `pkg/analysis_server/test`) cannot be run using `pub run test` which I believe is because:

a) `pub`/`pub run` does not support projects where the `.packages` file isn't at the same level as the `pubspec.yaml`. This also seems by design an not likely to change.
b) the `.packages` file for the SDK repo doesn't list `test`'s dependencies anyway, so they cannot be resolved (note: the deps are only required when running through the runner; eg. for `pub run` and not when run on `dart`, which is why it doesn't cause any issues there).

This makes it difficult for an editor with a test runner. If we always invoke tests via `pub run test` then tests in some projects (certainly the SDK, but I don't know if anyone else has a project structured this way) will fail. We can try and fall back to `dart` for these projects (and give a text output of the tests since there will be no JSON output) but knowing when to do that is also kind of complicated or involves trying `pub run test` and then detecting the failure. Explaining the issue to the user to let them pick would not be a great experience.

For now, I'm going to do my best to detect this (eg. by checking if the `.packages` is in the project root, not above it) but it seems like there should be a better way. It would be great if there was a standard way to run tests for a Dart project that "just works" and supports machine-parsable output.

**Comment by grouma** [comment]
I'll copy over some thoughts from the previous thread. In short, the issue is that @DanTup wants to use a feature of the test runner (JSON reporter) when none of the tests in the SDK use the test runner. I see that there are three potential solutions to this issue. None of them are ideal.

1. Update the `DEPS` and `.packages` files to include all the dependencies of `package:test`. This will allow tests to be executed as follows:

     `dart --packages=../../.packages package:test/src/executable.dart`

2. For packages you wish to run with the test runner, update their `.pubspec.yaml` so that they are valid and can successfully be called with `pub get`. This will require a number of `dependency_overrides` as there are several package dependencies that are not released.

3. Update `test.py` to use the test runner and forward any test configuration. This will of course require updating the `DEPS`. This will allow @DanTup to enable configurable test runners and make use of `test.py`.

Both option 1 and 2 will run the tests slightly different from what is currently advised. Option 3 seems like a non trivial amount of work. It's also conceivable that some tests are not compatible with the test runner today.

**Comment by DanTup** [comment]
Thanks! Though it's worth noting that all of the options above are specifically fixing this for the SDK repo; but since as far as I know it's setup is not unsupported, it's possibly (albeit unlikely?) that there are other codebases that also have this problem. I would be much more interested in a solution that would support all of them (although I used the SDK as an example because it's the first one I tried and it failed, I'm keen for my test runner to "just work" for as many people are possible without them having to change things or be set up in specific ways).

**Comment by matanlurey** [comment]
It's very unlikely this impacts traditional projects/packages.

The SDK was not built to be used with pub, much less pub run test.

**Comment by grouma** [comment]
Yeah it's unfortunate that it doesn't work with the SDK but I don't think it should be a blocker to providing test support in VS. As Matan mentioned, it is very unlikely that a package doesn't support `pub run test`. Moreover, our default Travis support goes through the test runner.

**Comment by matanlurey** [comment]
> Sure, but it's still a bit sucky that Dart's own tests using its own test framework can't be run in any of its editors test runners.

This is much much more meta issue and isn't reasonable to expect the `pkg/test` team to solve.

**Comment by DanTup** [comment]
Sure, I wasn't expecting `test` (or possibly even anyone) to solve it, I'm just trying to make sure I make the best call I can in Dart Code. If the decision is that codebases that are set up like this (*which may or may not be just the SDK - I'm under the impression that the SDK isn't doing anything specifically unsupported as far as Dart is concerned*) simply don't support the test runner, and that non-test-runner invocations (eg. via `dart`) will not support the JSON reporter, we should just be explicit about that. Right now I'm not sure whether this is a conscious or accidental decision (or even the latter that has/should become the former).

Right now the implementation of my test runner breaks the ability to run the SDK tests at all (I used to be able to press F5 and run tests and get a text output via `dart`, but now I'll just get an error message from `pub run test`). I must fix this somehow (it's not reasonable to not be able to debug these tests, even if I can't get JSON output) so I'm exploring the options and gathering information to figure out how best to do it.

If this will never be supported, that's fine (though a shame). I just then need to figure out the most reliable way to detect this scenario so I can pick the right debugger to avoid (or at least minimise) breaking people.

I don't feel I know enough about how all of these things (`pkg:test`, `.packages`, etc.) work to do the best thing without input from others.
